### PR TITLE
Add the project access token

### PR DESCRIPTION
**Pull request by randomswdev**
Add support for project access tokens (issue https://github.com/gitlabhq/terraform-provider-gitlab/issues/508).

The code is complete, but the pull request is in WIP becase it has a coupkle of prereqs:

- an updated version of the go-gitlab library. We already contributed the code and waiting for the review
- GitLab version 13.10, because, as discussed in the issue, version 13.9 has a bug in this area and it is planned to be fixed in 13.10

### CODE REVIEW

**Comment by randomswdev** [comment]
GitLab 13.10 has been release, so this prereq is now available. We are waiting for our puyll request fir the gitlab library to be accepted, before removing the WIP from this one.

In the meanwhile, is there anythong we have to do to have the test suite use Gitlab 13.10?

File: gitlab/resource_gitlab_project_access_token_test.go
Comment (PedDavid) [inline gitlab/resource_gitlab_project_access_token_test.go:204]: May be out of scope but I think it could be interesting to have a test case on the resource usage itself, e.g.: Adding the token to a CI/CD Variable and asserting it was correctly populated.

Other than that everything looks great :+1:

**Review by PedDavid — APPROVED** [review]

File: gitlab/resource_gitlab_project_access_token_test.go
Comment (randomswdev) [reply gitlab/resource_gitlab_project_access_token_test.go:204]: Not sure I understand you suggestion here.

Comment (PedDavid) [reply gitlab/resource_gitlab_project_access_token_test.go:204]: Having a test that uses the created resource itself (the project token), for example, in a project variable.

In practice should be as simple as a new test case where the config would have the following as well:

```
resource "gitlab_project_variable" "sample_project_variable" {
    project = gitlab_project.foo.id
    key = "sample_variable_key"
    value = gitlab_project_access_token.bar.token
}
```

Maybe this could be added to  `## Example Usage` in docs/resources/project_access_token.md as well

Comment (randomswdev) [reply gitlab/resource_gitlab_project_access_token_test.go:220]: I implemented the suggested change in my last commit.

Comment (PedDavid) [reply gitlab/resource_gitlab_project_access_token_test.go:220]: Thank you, this is great!

**Comment by Fuuzetsu** [comment]
Is there anything left to do here? This would be great to have and it was approved and is green in CI.

**Comment by nicholasklick** [comment]
@armsnyder mind reviewing this?

**Comment by nicholasklick** [comment]
@Fuuzetsu it needs to be approved and merged by a maintainer

**Comment by nicholasklick** [comment]
@mattkasa Mind reviewing this?

**Comment by davidholsgrove** [comment]
Hi, any chance this can be merged and a new cut of the terraform provider released? Would be great to have this resource in terraform

**Comment by randomswdev** [comment]
@nicholasklick @mattkasa @armsnyder I have rebased the change to resolve the conflict on the go.mod and go.sum files. Is there anythin missing on my side that prevents merging?

File: gitlab/resource_gitlab_project_access_token_test.go
Comment (mattkasa) [inline gitlab/resource_gitlab_project_access_token_test.go:244]:
```suggestion
    project   = gitlab_project.foo.id
    key       = "my_proj_access_token"
    value     = gitlab_project_access_token.bar.token
```

File: gitlab/resource_gitlab_project_access_token_test.go
Comment (mattkasa) [inline gitlab/resource_gitlab_project_access_token_test.go:197]:
```suggestion
  name = "my project token"
  project = gitlab_project.foo.id
  expires_at = "2022-04-01"
  scopes = ["read_repository" , "api", "write_repository", "read_api"]
```

File: gitlab/resource_gitlab_project_access_token_test.go
Comment (mattkasa) [inline gitlab/resource_gitlab_project_access_token_test.go:217]:
```suggestion
  name = "my new project token"
  project = gitlab_project.foo.id
  expires_at = "2022-05-01"
  scopes = ["api"]
```

File: gitlab/resource_gitlab_project_access_token_test.go
Comment (mattkasa) [inline gitlab/resource_gitlab_project_access_token_test.go:237]:
```suggestion
  name = "my new project token"
  project = gitlab_project.foo.id
  expires_at = "2022-05-01"
  scopes = ["api"]
```

File: docs/resources/project_access_token.md
Comment (mattkasa) [inline docs/resources/project_access_token.md:20]:
```suggestion
  project   = gitlab_project.example.id
  key       = "pat"
  value     = gitlab_project_access_token.example.token
```

**Review by mattkasa — COMMENTED** [review]
@randomswdev this looks really good, thank you for implementing this!

I left some `nitpick:` suggestions involving whitespace consistency, if you want to apply those or disregard, either is ok.

Either way, I'm happy to merge this once you've had time to take a look and let me know 👍

**Comment by randomswdev** [comment]
> @randomswdev this looks really good, thank you for implementing this!
> 
> I left some `nitpick:` suggestions involving whitespace consistency, if you want to apply those or disregard, either is ok.
> 
> Either way, I'm happy to merge this once you've had time to take a look and let me know 👍

@mattkasa I should have implemented all the requested changes. Thanks for reviewing this pull request.

**Comment by HaveFun83** [comment]
can someone have look here please? 
Too bad that this PR has not made it into the release

**Comment by mattkasa** [comment]
@randomswdev looks like acceptance tests are failing, can you please rebase this against `master` so it has the latest fixes?

File: gitlab/resource_gitlab_project_access_token.go
Comment (armsnyder) [inline gitlab/resource_gitlab_project_access_token.go:153]: This is likely a bug. `R004` is telling you that `projectAccessToken.Scopes` is not a valid type. What did the linter issue message say before you ignored it? See https://github.com/bflad/tfproviderlint/tree/main/passes/R004

File: gitlab/resource_gitlab_project_access_token.go
Comment (armsnyder) [inline gitlab/resource_gitlab_project_access_token.go:118]: The resource ID format should be `<project-id>:<token-id>`. A good rule of thumb is to include all IDs that you need to pass to the go-gitlab client to GET the object. You should also parse the project ID from the resource ID instead of reading the attribute. This will help you avoid needing a custom import function since you can then use `ImportStatePassthrough`. See `gitlab_project_mirror` for reference:

https://github.com/gitlabhq/terraform-provider-gitlab/blob/8e2faa78634fe687a5d491e3370ca9fd206f2476/gitlab/resource_gitlab_project_mirror.go#L18-L20

https://github.com/gitlabhq/terraform-provider-gitlab/blob/8e2faa78634fe687a5d491e3370ca9fd206f2476/gitlab/resource_gitlab_project_mirror.go#L132-L137

Please remember to add a `## Import` section to the docs as well. 😉

**Comment by randomswdev** [comment]
> @randomswdev looks like acceptance tests are failing, can you please rebase this against `master` so it has the latest fixes?

@mattkasa I rebased and this fixed ti the acceptance tests. I also made some minor changes to make the linter happy.
Now all the checks are ok.

File: gitlab/resource_gitlab_project_access_token.go
Comment (armsnyder) [inline gitlab/resource_gitlab_project_access_token.go:92]: Please include the resource ID in debug log messages

File: gitlab/resource_gitlab_project_access_token_test.go
Comment (armsnyder) [inline gitlab/resource_gitlab_project_access_token_test.go:106]: Nit-pick: Typo
```suggestion
		return fmt.Errorf("Project Access Token does not exist")
```

File: gitlab/resource_gitlab_project_access_token.go
Comment (armsnyder) [inline gitlab/resource_gitlab_project_access_token.go:49]: Nit-pick: This string-type validation should not be necessary since you've already specified the type in the schema, and the docs say Terraform guarantees that `i` is "guaranteed to be of the proper Schema type".
```suggestion
					v := i.(string)
```

File: gitlab/resource_gitlab_project_access_token_test.go
Comment (armsnyder) [inline gitlab/resource_gitlab_project_access_token_test.go:158]: You don't have anywhere in your tests where you test the behavior of your destroy function, so it's possible that tokens are not being deleted from GitLab. The `testAccCheckGitlabProjectAccessTokenDestroy` function is meant to validate that there are no more `gitlab_project_access_token`s for the project, not that the project is deleted. Depending on how comfortable you are with the test framework, it would be great if you could improve this. Here's an example where this is done right:

https://github.com/gitlabhq/terraform-provider-gitlab/blob/446a7d13f859235caccc58ab8a0129d89a61c04d/gitlab/resource_gitlab_project_approval_rule_test.go#L166-L174

**Review by armsnyder — CHANGES_REQUESTED** [review]
Thank you for the contribution, and for remembering to implement the `d.SetId("")` case! ❤️ 

In addition to my inline comments, can you fix the tfproviderlint issues?
1. Add an Import function please.
2. Handle errors with `Set`ting complex types.

File: gitlab/resource_gitlab_project_access_token.go
Comment (armsnyder) [reply gitlab/resource_gitlab_project_access_token.go:153]: I checked the GitHub Actions history and it looks like there was never a `R004` issue on this line. So, disregard!

**Comment by randomswdev** [comment]
We addessed all the requested changes. Please review the last commit.

File: gitlab/resource_gitlab_project_access_token.go
Comment (randomswdev) [reply gitlab/resource_gitlab_project_access_token.go:116]: No import added as described in the comment above

File: gitlab/resource_gitlab_project_access_token_test.go
Comment (randomswdev) [reply gitlab/resource_gitlab_project_access_token_test.go]: Done

File: gitlab/resource_gitlab_project_access_token.go
Comment (randomswdev) [reply gitlab/resource_gitlab_project_access_token.go]: Done

File: gitlab/resource_gitlab_project_access_token_test.go
Comment (randomswdev) [reply gitlab/resource_gitlab_project_access_token_test.go:186]: Implemented testing of the destroy function. Had to add a sleep because GitLab immediately returns from the destroy REST call but then deletes the token lazely. So we have to wait some time before checking if the token still exists.

File: gitlab/resource_gitlab_project_access_token.go
Comment (randomswdev) [reply gitlab/resource_gitlab_project_access_token.go]: Modified few log messages to include the relevant info.

**Comment by Fuuzetsu** [comment]
> > In addition to my inline comments, can you fix the tfproviderlint issues?
> > ```
> > 1. Add an Import function please.
> > ```
> 
> @armsnyder the import functionality does not really make sense, because the API do not allow to read the token. So, if terraform creates a token, in the creation response, we receive the token value and can save it in the state in order to use it elsewhere. But if we execute an import, we cannot retrieve the token value; so after importing the token we cannot really use it because we are unable to retrieve the token value. So importing it is somehow useless.

Just a nitpick: I think if you have made the token manually and you subsequently want to put it in terraform state, you should be able to do an import from a file for example. Or something. Just anything to allow getting manual token into the state.

Anyway, should not be a blocker for this in my opinion (I'm not affiliated with the project)

**Comment by ThePlenkov** [comment]
I would really like to have this feature available. Our company restricts scopes of deploy tokens to read-only. So only access tokens can allow us to push from pipeline..

**Comment by randomswdev** [comment]
@armsnyder @mattkasa @PedDavid I think, based on the release cadence of this provider, that a new relase tmeframe could be approaching. I please ask you to perform a final review and integrate this pull request, so that it can make it in the next release.

**Comment by armsnyder** [comment]
Hi @randomswdev, this PR looks really good and I'd also like it in the next release. I have one more piece of feedback: Can you change the resource id format from `<token-id>` to `<project-id>:<token-id>`? This will make it easier to add an importer and also aligns to the id format of our other project resource ids.

**Comment by Dutchy-** [comment]
I don't know when the next release is, exactly, but I would like to know if this feature is going to make it in

**Comment by pragone** [comment]
I just thought I'd add my voice here so it's included in the following release. It seems this PR is looking good, and I (and my team) would really appreciate having it.

**Comment by randomswdev** [comment]
> 
> 
> Hi @randomswdev, this PR looks really good and I'd also like it in the next release. I have one more piece of feedback: Can you change the resource id format from `<token-id>` to `<project-id>:<token-id>`? This will make it easier to add an importer and also aligns to the id format of our other project resource ids.

@armsnyder is this change a strong requirement for having this pull request approved? If so I'll try to implement the change. I would like to have this pull request merged as sson as possible, considering it is open since March.

**Comment by randomswdev** [comment]
> Hi @randomswdev, this PR looks really good and I'd also like it in the next release. I have one more piece of feedback: Can you change the resource id format from `<token-id>` to `<project-id>:<token-id>`? This will make it easier to add an importer and also aligns to the id format of our other project resource ids.

@armsnyder I just implemented the requested change. I hope this merge request can be included i nthe next release of the provider.

**Comment by erSitzt** [comment]
Any news on when the next release is coming and if this MR will be included ?
Would be great ! :)
 @armsnyder

**Comment by nagyv** [comment]
@roidelapluie As @armsnyder does not have much availability, would you mind providing a review here? Based on https://github.com/gitlabhq/terraform-provider-gitlab/pull/588#issuecomment-976297086 his request is fulfilled.

File: gitlab/resource_gitlab_project_access_token.go
Comment (timofurrer) [inline gitlab/resource_gitlab_project_access_token.go:15]: Please implement the `Importer` instead of disabling [`XR002`](https://github.com/bflad/tfproviderlint/blob/main/xpasses/XR002/README.md)

File: gitlab/resource_gitlab_project_access_token.go
Comment (timofurrer) [inline gitlab/resource_gitlab_project_access_token.go:23]: The `project` field is of type `schema.TypeString` for (almost) all other resources. I highly recommend to change it from an int to string here, too.

**Review by timofurrer — CHANGES_REQUESTED** [review]
This change set also requires a rebase to the current `master` in order to upgrade to SDK v2.

File: gitlab/resource_gitlab_project_access_token.go
Comment (timofurrer) [inline gitlab/resource_gitlab_project_access_token.go:17]: Please use the context-aware CRUD functions, see the [upgrade guide](https://www.terraform.io/plugin/sdkv2/guides/v2-upgrade-guide#more-support-for-context-context).

Also make sure to use the new `ctx` to pass it to the `gitlab` API functions as `gitlab.WithContext(ctx)`. It's usually the last parameter of the API functions.

**Review by timofurrer — CHANGES_REQUESTED** [review]

**Comment by randomswdev** [comment]
> This change set also requires a rebase to the current `master` in order to upgrade to SDK v2.

@timofurrer I should have addressed all of your change requests. Let me know ifg it is fine and if it can be integrated.

File: gitlab/resource_gitlab_project_access_token.go
Comment (randomswdev) [reply gitlab/resource_gitlab_project_access_token.go]: Resolved in last commit

**Review by timofurrer — APPROVED** [review]

**Comment by randomswdev** [comment]
@timofurrer the tests failed. I just committed a fix.

File: gitlab/resource_gitlab_project_access_token.go
Comment (armsnyder) [reply gitlab/resource_gitlab_project_access_token.go:160]: I'm worried about this again, since I've seen there have been breaking changes in the go-gitlab package around slices. Can you resolve the `XR004` issue and remove the `//lintignore` comment? (You can resolve it by handling the error return value.)

File: gitlab/resource_gitlab_project_access_token.go
Comment (armsnyder) [inline gitlab/resource_gitlab_project_access_token.go:17]: This resource should really have an importer since it can become deleted from the Terraform state if the PAT is not found.

File: gitlab/resource_gitlab_project_access_token_test.go
Comment (armsnyder) [inline gitlab/resource_gitlab_project_access_token_test.go:89]: This is not great. Can you use `gomega.Eventually` here?

Something like

```go
import (
	. "github.com/onsi/gomega"
)

...

return func(s *terraform.State) error {
	return InterceptGomegaFailure(func() {
		Eventually(func() error {
			// Your test code here.
		}).Should(Succeed())
...
```

See the `testAccCheckGitlabProjectApprovalRuleAttributes` for example: https://github.com/gitlabhq/terraform-provider-gitlab/blob/79a17cbecec558f274a0d9a32dc22bdeff5d0750/gitlab/resource_gitlab_project_approval_rule_test.go#L92

**Review by armsnyder — CHANGES_REQUESTED** [review]

**Comment by armsnyder** [comment]
I added some more feedback. I know a lot of folks are waiting on this, so we could always merge this and open new issues to circle back on.

File: gitlab/resource_gitlab_project_access_token.go
Comment (randomswdev) [reply gitlab/resource_gitlab_project_access_token.go]: Modified the code to remove the XR004

File: gitlab/resource_gitlab_project_access_token.go
Comment (randomswdev) [reply gitlab/resource_gitlab_project_access_token.go:17]: It is very complex to add an importer here, because GitLab provides the token value only when the token is created. Importing the token thus means finding a way for providing also the token value to the importer.

The same behaviour is currently implemented in

https://github.com/gitlabhq/terraform-provider-gitlab/blob/8bc5606d786d73cb9e1947d0f9bc941c2e4f61b0/gitlab/resource_gitlab_deploy_token.go#L18

I wouyld skip implementing the importer for this type of resource.

**Comment by randomswdev** [comment]
@armsnyder I addressed ypur concerns, excluding the importer related one (see my comment above).

File: gitlab/resource_gitlab_project_access_token.go
Comment (armsnyder) [reply gitlab/resource_gitlab_project_access_token.go:17]: If we did have an importer we would not set the `token` field. The purpose would be to track the resource, for example so a user could clean up the token with `terraform destroy`. But this is not a high-value use-case, so I don't mind having no importer for now.

File: gitlab/resource_gitlab_project_access_token_test.go
Comment (randomswdev) [reply gitlab/resource_gitlab_project_access_token_test.go]: Sorry, I missed the example. Updated the code as suggested.

**Review by armsnyder — APPROVED** [review]
lgtm

**Pull request merged**
